feat(cli): framework for interactive prompts in cli

ren-yamanashi opened the pull request:
Fixes #4086 

## Changes description

Implemented a feature that allows selection of javascript tools (package manager / linter / formatter / test tool) when the `npx projen new` command is executed.

## Confirmation of operation

- [x] Pass unit tests.
- [x] Confirm that the functionality works correctly by actually executing the command.

> Confirm that the functionality works correctly by actually executing the command.

https://github.com/user-attachments/assets/c4607a22-3379-46db-adec-5f55aba49a86


---
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ren-yamanashi committed:
feat(cli): framework for interactive prompts in cli

ren-yamanashi committed:
feat: update

ren-yamanashi commented:
MEMO: 

To make biome etc. available as a linter, you need to modify `. /.jsii` needs to be modified 
(currently biome is not added to .jsii, so biome cannot be selected as a linter)

ren-yamanashi committed:
feat: update

ren-yamanashi committed:
chore: update package.json

ren-yamanashi committed:
chore: update .projenrc.ts

ren-yamanashi committed:
test: update

ren-yamanashi committed:
test: update

ren-yamanashi committed:
test: update

ren-yamanashi committed:
feat: add interactive CLI prompts for CI environment

ren-yamanashi committed:
feat: add consola as a bundled dependency and update CLI prompts for test environment

ren-yamanashi committed:
fix: update project type detection and enable ESLint, Jest, and Prettier by default in snapshots

ren-yamanashi committed:
refactor: remove unused CLI prompts mock for project creation tests

ren-yamanashi committed:
feat: update selectJsTools to return undefined instead of null and add tests for user prompts

ren-yamanashi commented:
### TODO

- [ ] package name
- [ ] package manager
- [ ] create issue
  - add vitest
  - update .jsii (add biome)

ren-yamanashi committed:
feat: refactor interactive CLI prompts and consolidate selectJsTools functionality

ren-yamanashi committed:
feat: update project creation options to include package manager and refactor snapshot tests

ren-yamanashi commented:
https://github.com/user-attachments/assets/e5e10611-f6aa-4ac9-b3a3-d08fe7363963

ren-yamanashi committed:
refactor: remove export modifier from SelectItemValue type

ren-yamanashi committed:
feat: enhance selectJsTools to support optional project options and update tests for improved user prompts

ren-yamanashi committed:
refactor: add missing GIVEN comments in test cases for clarity

ren-yamanashi committed:
Merge branch 'main' of https://github.com/ren-yamanashi/projen into feat/4086-interactive-prompt-cli

mrgrain commented on src/projects.ts:
```diff
@@ -107,13 +108,29 @@ function createProject(opts: CreateProjectOptions) {
+    args: jsTools
+      ? {
+          ...opts.projectOptions,
+          name: jsTools.projectName,
+          eslint: opts.projectOptions.eslint || jsTools.linter === "eslint",
+          prettier:
+            opts.projectOptions.prettier || jsTools.formatter === "prettier",
+          jest: opts.projectOptions.jest || jsTools.testTool === "jest",
+          packageManager:
+            opts.projectOptions.packageManager || jsTools.packageManager,
+        }
```
That's fine. Could you raise an issue for me to investigate?

ren-yamanashi replied:
Upon checking, it seems this issue has been resolved. Therefore, I have also updated `biome` to be selectable.

see: https://github.com/projen/projen/blob/b918a53ffb4110b15602ec1b024218a4ca6e4493/src/cli/prompts/operations/select-js-tools.ts

mrgrain reviewed (changes requested):
Thanks for getting this started! Two design issues I have with this approach:

- We should not change `createProject` to include the interaction. The interaction should be a CLI only feature and its result passed on to the existing `createProject`. This will also allow you to revert all the test changes you had to make. **This is a blocking requirement.**
- I'm okay with the current approach of hardcoded, inlined prompting in the CLI. However next step should be to come up with an extensible way to declare prompts as part of the inventory and have the CLI discover the prompts fro the inventory. This should be extensible for external packages as well. **We probably want a small RFC for this, so let's not make this part of this PR.**

Otherwise, we are missing a flag to enable/disable interactive mode, e.g. `--interactive` and `--no-interactive`. We should also auto-detect non TTY and CI environments (the latter using `process.env.CI`) and disable interaction automatically.

ren-yamanashi committed:
Merge branch 'main' of https://github.com/ren-yamanashi/projen into feat/4086-interactive-prompt-cli

ren-yamanashi committed:
chore: revert unrelated changes

ren-yamanashi committed:
chore: rm changes

ren-yamanashi committed:
chore: rm changes

ren-yamanashi committed:
chore: rm comments

ren-yamanashi committed:
chore: rm changes

ren-yamanashi committed:
chore: rm changes

ren-yamanashi commented on src/run-projenrc-json.task.ts:
```diff
@@ -24,6 +24,8 @@ if (!type) {
 }
 
+// top-level await is not supported in this file
+// eslint-disable-next-line @typescript-eslint/no-floating-promises
```
https://github.com/projen/projen/pull/4326/commits/fb0180d7fecd6b9a759800e0a9f994125f954ce5

ren-yamanashi commented on src/projects.ts:
```diff
@@ -69,8 +70,8 @@ export class Projects {
    */
-  public static createProject(options: CreateProjectOptions) {
-    createProject(options);
+  public static async createProject(options: CreateProjectOptions) {
```
https://github.com/projen/projen/pull/4326/commits/fb55bbd25544ffcc4424f77b2314736fc19d0e49

ren-yamanashi committed:
chore: rm comment outs

ren-yamanashi committed:
test: update

ren-yamanashi committed:
feat: support interactive mode

ren-yamanashi commented on test/util.ts:
```diff
@@ -99,42 +99,68 @@ export function synthSnapshotWithPost(project: Project) {
+  const outdir = options.tmpdir ?? mkdtemp();
+
+  try {
+    const projectdir = setupProjectDir(outdir, options);
```
Reverted this changes so, I resolve conversation.

ren-yamanashi committed:
test: update

ren-yamanashi commented:
@mrgrain 

> * We should not change `createProject` to include the interaction. The interaction should be a CLI only feature and its result passed on to the existing `createProject`. This will also allow you to revert all the test changes you had to make. **This is a blocking requirement.**


Thanks review!

I added `interactive` cli option, making interactive functionality a cli only feature.([see](https://github.com/projen/projen/blob/b918a53ffb4110b15602ec1b024218a4ca6e4493/src/cli/cmds/new.ts#L62-L66))

And I fixed `createProject` method was concerned with interactive functionality.([see this comment](https://github.com/projen/projen/pull/4326#discussion_r2441586235))

Commit:
Merge branch 'main' of https://github.com/ren-yamanashi/projen into feat/4086-interactive-prompt-cli

Commit:
chore: resolve conflicts

Commit:
chore: rm changes

Commit:
chore: rm changes

Commit:
chore: rm changes

Commit:
chore: rm changes

mrgrain committed:
Merge branch 'main' into feat/4086-interactive-prompt-cli

mrgrain commented:
This lists too many yarn versions. Order alphabetically.
<img width="209" height="145" alt="image" src="https://github.com/user-attachments/assets/653e3c98-1022-4ef3-89fd-5c450aee30b4" />

mrgrain commented:
`Ctrl+C` on selection prompt produces an unsightly error.
<img width="875" height="160" alt="image" src="https://github.com/user-attachments/assets/d4db539e-6d2d-4b90-816a-2253bc25dea9" />

Commit:
update interactive-cli-prompt

Commit:
process exit when cancelled prompt

Commit:
rm changes

Commit:
test update

Commit:
Merge branch 'main' of https://github.com/ren-yamanashi/projen into feat/4086-interactive-prompt-cli

Commit:
fix lint error

Commit:
fix test

ren-yamanashi commented:
@mrgrain 

I fixed code so that process exit when `Ctrl+C` is pressed.
And also sorted items alphabetically.

https://github.com/user-attachments/assets/d9dedf6a-95e1-4420-8a25-90a7c78b7c3f

mrgrain commented on src/cli/prompts/core/interactive-cli-prompt.ts:
```diff
@@ -0,0 +1,115 @@
+      switch (args.cancel) {
+        case "exit": {
+          // when user cancels the prompt, exit the process with code 1
+          process.exit(1);
```
You might want to refactor this into a function so it can be reused in the various places.
